internal/provider/cloudinit_config.go

```diff
@@ -0,0 +1,173 @@
+package provider
+
+import (
+	"bytes"
+	"compress/gzip"
+	"context"
+	"encoding/base64"
+	"fmt"
+	"io"
+	"mime/multipart"
+	"net/textproto"
+	"strconv"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+	"github.com/hashicorp/terraform-plugin-framework/path"
+	"github.com/hashicorp/terraform-plugin-framework/types"
+	"github.com/hashicorp/terraform-plugin-log/tflog"
+	"github.com/hashicorp/terraform-provider-cloudinit/internal/hashcode"
+)
+
+// Model and functionality of data source and resource are equivalent.
+type configModel struct {
+	ID           types.String      `tfsdk:"id"`
+	Parts        []configPartModel `tfsdk:"part"`
+	Gzip         types.Bool        `tfsdk:"gzip"`
+	Base64Encode types.Bool        `tfsdk:"base64_encode"`
+	Boundary     types.String      `tfsdk:"boundary"`
+	Rendered     types.String      `tfsdk:"rendered"`
+}
+
+type configPartModel struct {
+	ContentType types.String `tfsdk:"content_type"`
+	Content     types.String `tfsdk:"content"`
+	FileName    types.String `tfsdk:"filename"`
+	MergeType   types.String `tfsdk:"merge_type"`
+}
+
+func validateConfigModel(config configModel) diag.Diagnostics {
```

Comment:
Super nit: Here and below seem like they could be methods on the `configModel` type, rather than package functions that take it in as a parameter, e.g.

```go
func (c configModel) validate() diag.Diagnostics
func (c *configModel) update() diag.Diagnostics
func (c *configModel) render(ctx context.Context) diag.Diagnostics 
```

etc.

Reply:
good call, I updated the code to utilize receiver methods instead. I only left `renderPartsToWriter` as it felt more "outside" the configModel, but I can update that one as well if you think it'd make more sense